activerecord/lib/active_record/future_result.rb

```diff
@@ -25,7 +25,7 @@ def schedule!(session)
     end
 
     def execute!(connection)
-      execute_query(connection)
+      execute_query(connection, async: true)
```

Comment:
That doesn't sound right to me. `execute!` is called when the query can't be scheduled: https://github.com/rails/rails/blob/e9de0f51c60ec206ed0a1d2c143ec952a55814d6/activerecord/lib/active_record/connection_adapters/abstract/database_statements.rb#L540-L543

That's likely why you don't get the `ASYNC` in your logs.

What adapter do you use? And how do you test the query? Right now on master async queries are disabled unless you are inside a request/job cycle.

Reply:
It's mysql2 and it's a test that hits the index endpoint tested against #41372

Reply:
Hum, weird. I'll take a look.

Reply:
Checkout https://github.com/eileencodes/parallel_queries, create the dbs and then run `bundle exec ruby -Ilib:test test/controllers/dogs_controller_test.rb -n test_should_get_index` while tailing the logs.

Reply:
Oh don't forget to update the rails path, right now it's linking to my local copy.

Reply:
Ah, it's in test? That's likely why. I presume transactional fixtures are enabled, which means you are inside a transaction which prevent scheduling the query on a pool. Let me double check.

Reply:
Yes, I confirm, if you set `self.use_transactional_tests = false` in `DogsControllerTest`, you'll see ASYNC queries:

```
$ bundle exec ruby -Ilib:test test/controllers/dogs_controller_test.rb -n test_should_get_index
Run options: -n test_should_get_index --seed 31392

# Running:

D, [2021-02-11T16:45:36.353585 #79892] DEBUG -- :   Dog Load (0.3ms)  SELECT `dogs`.* FROM `dogs` WHERE `dogs`.`name` = 'fido 1'
D, [2021-02-11T16:45:36.355451 #79892] DEBUG -- :   Dog Load (1.1ms)  SELECT `dogs`.* FROM `dogs` WHERE `dogs`.`name` = 'fido 2'
D, [2021-02-11T16:45:36.356247 #79892] DEBUG -- :   Dog Load (0.2ms)  SELECT `dogs`.* FROM `dogs` WHERE `dogs`.`name` = 'fido 3'
D, [2021-02-11T16:45:36.357815 #79892] DEBUG -- :   Dog Load (0.2ms)  SELECT `dogs`.* FROM `dogs` WHERE `dogs`.`name` = 'fido 4'
D, [2021-02-11T16:45:36.358364 #79892] DEBUG -- :   Dog Load (0.2ms)  SELECT `dogs`.* FROM `dogs` WHERE `dogs`.`name` = 'fido 5'
D, [2021-02-11T16:45:36.360089 #79892] DEBUG -- :   Dog Load (0.2ms)  SELECT `dogs`.* FROM `dogs` WHERE `dogs`.`name` = 'fido 6'
D, [2021-02-11T16:45:36.360828 #79892] DEBUG -- :   Dog Load (0.2ms)  SELECT `dogs`.* FROM `dogs` WHERE `dogs`.`name` = 'fido 7'
D, [2021-02-11T16:45:36.361805 #79892] DEBUG -- :   Dog Load (0.2ms)  SELECT `dogs`.* FROM `dogs` WHERE `dogs`.`name` = 'fido 8'
D, [2021-02-11T16:45:36.362361 #79892] DEBUG -- :   Dog Load (0.2ms)  SELECT `dogs`.* FROM `dogs` WHERE `dogs`.`name` = 'fido 9'
D, [2021-02-11T16:45:36.363470 #79892] DEBUG -- :   Dog Load (0.2ms)  SELECT `dogs`.* FROM `dogs` WHERE `dogs`.`name` = 'fido 10'
D, [2021-02-11T16:45:36.371393 #79892] DEBUG -- :   ASYNC  (1.8ms)  SELECT * FROM dogs where name = 'fido 2'
D, [2021-02-11T16:45:36.371868 #79892] DEBUG -- :   ASYNC  (4.0ms)  SELECT * FROM dogs where name = 'fido 1'
D, [2021-02-11T16:45:36.378731 #79892] DEBUG -- :   ASYNC  (6.6ms)  SELECT * FROM dogs where name = 'fido 4'
D, [2021-02-11T16:45:36.379211 #79892] DEBUG -- :   ASYNC  (6.2ms)  SELECT * FROM dogs where name = 'fido 3'
D, [2021-02-11T16:45:36.379444 #79892] DEBUG -- :   ASYNC  (6.3ms)  SELECT * FROM dogs where name = 'fido 5'
D, [2021-02-11T16:45:36.379880 #79892] DEBUG -- :   Dog Load (1.2ms)  SELECT `dogs`.* FROM `dogs`
D, [2021-02-11T16:45:36.379660 #79892] DEBUG -- :   ASYNC  (6.5ms)  SELECT * FROM dogs where name = 'fido 6'
D, [2021-02-11T16:45:36.383434 #79892] DEBUG -- :   ASYNC  (2.6ms)  SELECT * FROM dogs where name = 'fido 7'
D, [2021-02-11T16:45:36.383829 #79892] DEBUG -- :   ASYNC  (2.9ms)  SELECT * FROM dogs where name = 'fido 8'
D, [2021-02-11T16:45:36.384344 #79892] DEBUG -- :   ASYNC  (0.6ms)  SELECT * FROM dogs where name = 'fido 10'
D, [2021-02-11T16:45:36.385191 #79892] DEBUG -- :   ASYNC  (0.9ms)  SELECT * FROM dogs where name = 'fido 9'
```

Reply:
That feels like a pretty confusing user experience. We ask for async, it says it's async if you're debugging the args passed around and ultimately it returns a future result but it wasn't actually async or future. In the case where it's not actually async then shouldn't it go through the non-async execution to avoid confusion?

Reply:
I think this is still incorrect when linked to #41372 though. The logs should say `ASYNC Dog Load`.

```
D, [2021-02-11T16:45:36.353585 #79892] DEBUG -- :   Dog Load (0.3ms)  SELECT `dogs`.* FROM `dogs` WHERE `dogs`.`name` = 'fido 1'
D, [2021-02-11T16:45:36.355451 #79892] DEBUG -- :   Dog Load (1.1ms)  SELECT `dogs`.* FROM `dogs` WHERE `dogs`.`name` = 'fido 2'
D, [2021-02-11T16:45:36.356247 #79892] DEBUG -- :   Dog Load (0.2ms)  SELECT `dogs`.* FROM `dogs` WHERE `dogs`.`name` = 'fido 3'
D, [2021-02-11T16:45:36.357815 #79892] DEBUG -- :   Dog Load (0.2ms)  SELECT `dogs`.* FROM `dogs` WHERE `dogs`.`name` = 'fido 4'
```

Reply:
> We ask for async, it says it's async if you're debugging the args

Well, the argument says we'd like async, if that's not possible, we degrade to sync.

> ultimately it returns a future result but it wasn't actually async or future

Since it's a fallback, I chose to use the same flow so that things such as `:after_initialize` callbacks would be triggered at a similar time. The goal was to be as close to the `async` behavior as possible. That's pretty much what we're discussing here: https://github.com/rails/rails/pull/41372#discussion_r572309650 with @matthewd

Reply:
> The logs should say `ASYNC Dog Load`.

No because you don't let the executor the time to pop the workloads:

```ruby
    (1..10).each do |i|
      Dog.where(name: "fido #{i}").load_async.to_a
    end
```

You immediately call `to_a` on the relation. Try:

```ruby
    dog_queries = (1..10).map do |i|
      Dog.where(name: "fido #{i}").load_async
    end
   dog_queries.each(&:to_a)
```

Reply:
Ok, I see. I still think this behavior is surprising from a user experience perspective. This is going to be a common mistake users make when implementing async queries and I think we're gonna get a lot of bug reports where the answer is "you're doing it wrong". 

> That's pretty much what we're discussing here: #41372 (comment) with @matthewd

I totally missed this entire conversation previously, I've started using Hey for oss emails and I'm not quite used to how it works with threading. I kind of disagree with Matthew here, but I'll comment on the other PR and close this one so we don't have diverging conversations.

Reply:
👌